node/src/Pos/Explorer/DB.hs

```diff
@@ -104,10 +133,115 @@ prepareExplorerDB = do
+    convertOldEpochBlocksFormat =
+        runConduitRes  $ epochsSource
+                      .| epochPagesConduit
+                      .| epochPagesExplorerOpSink
+      where
+        -- 'Source' corresponding to the old @Epoch@ format that contained all
+        -- 21600 @[HeaderHash]@ in a single @Epoch@ (in production).
+        epochsSource :: (MonadDBRead m) => Source (ResourceT m) (Epoch, [HeaderHash])
+        epochsSource = dbIterSource GStateDB (Proxy @EpochsIter)
+
+        -- 'Conduit' to turn the old @Epoch@ format that contained all
+        -- 21600 @[HeaderHash]@ in a single @Epoch@ to a new format that contains
+        -- @Epoch@ *and* @Page@ as a key since there are too many @[HeaderHash]@ to
+        -- load - performance reasons.
+        epochPagesConduit
+            :: (MonadDBRead m)
+            => Conduit (Epoch, [HeaderHash]) m (Map EpochPagedBlocksKey [HeaderHash])
+        epochPagesConduit = CL.map convertToPagedMap
+          where
+            convertToPagedMap
+                :: (Epoch, [HeaderHash])
+                -> Map EpochPagedBlocksKey [HeaderHash]
+            convertToPagedMap ehh = fromListWith (++) (convertToPaged ehh)
+
+            convertToPaged :: (Epoch, [HeaderHash]) -> [((Epoch, Page), [HeaderHash])]
+            convertToPaged (epoch, headerHashes) =
+                map convertHHsToEpochPages convertHHsToPages
+              where
+                convertHHsToEpochPages
+                    :: (Page, [HeaderHash])
+                    -> ((Epoch, Page), [HeaderHash])
+                convertHHsToEpochPages (page, headerHashes') =
+                    ((epoch, page), headerHashes')
+
+                -- | Take that huge chunk of @HeaderHash@es and page it.
+                convertHHsToPages :: [(Page, [HeaderHash])]
+                convertHHsToPages = zip [1..] $ splitEvery pageSize headerHashes
+                  where
+                    -- | TODO (ks): It's getting repeated all over the place.
+                    pageSize :: Int
+                    pageSize = 10
+
+                    splitEvery :: Int -> [a] -> [[a]]
+                    splitEvery _ [] = []
+                    splitEvery n xs = as : splitEvery n bs
+                      where
+                        (as,bs) = splitAt n xs
+
+        -- | Finally, we persist the map with the new format.
+        epochPagesExplorerOpSink
+            :: (MonadDB m)
+            => Sink (Map EpochPagedBlocksKey [HeaderHash]) m ()
+        epochPagesExplorerOpSink = CL.mapM_ persistEpochBlocks
+          where
+            -- | Persist atomically, all the operations together.
+            persistEpochBlocks
+                :: (MonadDB m)
+                => Map EpochPagedBlocksKey [HeaderHash]
+                -> m ()
+            persistEpochBlocks mapEpochPagedHHs =
+                -- Here we persist all the changes atomically.
+                writeBatchGState $ persistEpochPageBlocks ++ persistMaxPageNumbers
```

Comment:
As far as I understand, you still perform all batch writes separately. However, apparently this cannot be easily fixed, unless we want to load the whole list of operations into memory. We don't support transactions in RocksDB (yet), so we can't open a transaction, and concatenating all operations to produce one huge list would 1) probably occupy a ton of memory and 2) would render Conduit unneeded. I'd recommend keeping the code as-is, but we'd probably be better off hearing some advice from @pva701 or @gromakovsky.

Reply:
We have atomic persistence of an *Epoch* but not the persistence of multiple *Epoch*s.
As you said yourself, we would occupy too much memory if we drop conduit, and these key-value data stores aren't that flexible (in my experience) as relational are.
If we had a relational (say Oracle) we could post it on the database (in batches) and in the end commit all of it atomically.

I have an idea how to do it in RocksDB, but it's a hack and I don't want to do it that way.
I'll write about it here so if someone wants to be a hero, he can.
I'll write about the specific case, generalize away.
In this example, we have a (new) key `"e/epochs/" <> serialize' epoch <> "/" <> encodedPage`.

We could temporarily create a new key, say `"e/posting/epochs/" <> serialize' epoch <> "/" <> encodedPage` for the persistence of the new epochs using conduit.
We then add another one, say `"e/postingStatus/epochs/" <> serialize' epoch` with a simple `Bool` that designates whether all the keys are persisted (we could simplify this, but nevermind).
In the end, we simply check whether all the keys are `True` and if yes, we either _change their keys_ or save them under the original key again, delete the temp keys we created and voila - we have a very crude hackish atomic transaction.

Again, this seems (a bit) insane and I would rather leave it as-is.

Reply:
Whether a DB is relational or not is independent of whether it supports transactions. In fact, RocksDB [_does_ support transactions](https://github.com/facebook/rocksdb/wiki/Transactions), but AFAIU we need to use a different way of initializing the DB (called TransactionDB in RocksDB lingo) to make use of that.